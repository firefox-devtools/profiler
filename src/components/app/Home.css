/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.home {
  display: flex;
  width: 100%;
  min-height: 100vh;
  align-items: center;
  align-self: flex-start;
  justify-content: center;
  line-height: 1.5;
}

.homeSpecialMessage {
  padding: 8px 16px;
  border: 1px solid #000;
  border-radius: 3px;
  margin: 17px 0;
  background-color: #798fc8;
  color: #fff;
}

.homeSection {
  position: relative;
  overflow: hidden;
  width: 75%;
  max-width: 1200px;
  box-sizing: border-box;
  padding: 4em 8em;
  border: 1px solid #ccc;
  border-radius: 3px;
  background-color: #fff;
  box-shadow: 0 5px 25px #0b1f50;
  font-size: 130%;
}

.homeSection kbd {
  display: inline-block;
  padding: 0 0.5em;
  border: 1px solid #ccc;
  border-radius: 0.2em;
  margin: 0 0.2em;
  background-color: #f6f6f6;
  box-shadow: 0.1em 0.1em 0 #bbb;
}

.homeSectionItems {
  margin: 1em 0;
}

.homeSectionScreenshot {
  width: 390px;
  max-width: 100%;
  height: auto;
  border: 1px solid #d1d1d1;
  border-radius: 0.4em;
  box-shadow: 0 2px 4px #ddd;
  place-self: center;
}

.homeAdditionalContentTitle {
  margin-top: 0;
  margin-bottom: 10px;
  grid-column: 1 / -1; /* from start to end */
}

.homeActions > p:first-child,
.homeRecentUploadedRecordingsTitle {
  margin-top: 0;
}

.homeSectionButton {
  display: block;
  width: 100%;
  box-sizing: border-box;
  padding: 0.5em;
  border: 2px solid currentcolor;
  margin: 15px 0;
  background-color: #fff;
  color: var(--grey-50);
  font-size: 16px;
  font-weight: bold;
  line-height: 1.5;
  text-align: center;
  text-decoration: none;
  transition:
    background-color 250ms var(--animation-curve),
    color 250ms var(--animation-curve);
}

.homeSectionButton:not(:disabled) {
  color: var(--blue-60);
  cursor: pointer;
}

.homeSectionLoadProfile {
  position: relative;
}

.homeSectionActionButtons {
  display: flex;
}

.homeSectionActionButtons .homeSectionButton {
  flex: 1;
  margin: 5px 0;
}

.homeSectionActionButtons button:not(:first-of-type) {
  border-left: none;
}

.homeSectionLoadFromUrl {
  position: absolute;
  display: flex;
  width: 100%;
  justify-content: space-between;
}

.homeSectionActionButtons .homeSectionButton,
.homeSectionLoadFromUrl .homeSectionButton {
  padding: 0.3em 1em;
  font-size: 13px;
}

.homeSectionButton:not(:disabled):hover,
.homeSectionButton:not(:disabled)[aria-expanded='true'] {
  background-color: var(--blue-60);
  color: #fff;
}

.homeSectionPlus {
  position: relative;
  top: 2px;
  padding-right: 12px;
  padding-left: 5px;
  font-size: 26px;
  line-height: 0;
}

.homeSectionDocsIcon {
  position: relative;
  display: inline-block;
  width: 20px;
  height: 20px;
  margin: 0 10px -4px 3px;
  background: url(../../../res/img/svg/help-blue.svg);
  background-size: 100% 100%;
}

.homeSectionButton:hover .homeSectionDocsIcon {
  /* Make the icon white */
  filter: grayscale(1) brightness(3);
}

.homeInstructions,
.homeAdditionalContent {
  display: grid;
  margin-top: 30px;
  column-gap: 30px;

  /* minmax makes the grid ignore the element's intrinsic size.
   * If we use 1fr directly, the min is "auto", which depends on the content. If
   * the content is too large then this can break the layout. */
  grid-template-columns: minmax(0, 1fr) minmax(0, 1fr);
}

.homeInstructionsTransitionGroup {
  position: relative;
}

/* Only show the home transition when reduced motion is not preferred */
@media (prefers-reduced-motion: no-preference) {
  .homeTransition-enter {
    opacity: 0.1;
    transform: translateX(100px);
  }

  .homeTransition-enter.homeTransition-enter-active {
    opacity: 1;
    transform: translateX(0);
    transition:
      opacity 300ms ease-out,
      transform 300ms ease-out;
  }
}

.homeSectionUploadFromFileInput {
  display: none;
}

.homeSectionLoadFromUrlInput {
  flex: auto;
  margin: 0 10px 0 0;
}

.homeSectionLoadFromUrlSubmitButton {
  margin: 0;
}

@media (max-width: 1300px) {
  .homeSection {
    padding: 5em;
  }
}

@media (max-width: 1100px) {
  .homeSection {
    max-width: 80%;
    padding: 3em;
  }

  .homeAdditionalContent,
  .homeInstructions {
    /* we switch to use just one column instead of 2.
     * Note tha/ minmax makes the grid ignore the element's intrinsic size. */
    grid-template-columns: minmax(0, 1fr);
  }
}

@media (max-width: 767px) {
  .homeSection {
    width: 90%;
    max-width: 90%;
    padding-right: 2em;
    padding-left: 2em;
  }
}

@media (max-width: 499px) {
  .homeSection {
    width: 95%;
    max-width: 95%;
    padding-right: 1em;
    padding-left: 1em;
  }
}
