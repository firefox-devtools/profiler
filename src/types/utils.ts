/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/**
 * This type serves as documentation for how an array is meant to be used, but does
 * not support type checking. We often use an Array instead of a Map to translate
 * one type of index into another type of index. This is similar to how we use the
 * Map<K,V> type, but with the Array.
 */
export type IndexedArray<_IndexType, Value> = Array<Value>;

export type MixedObject = { [key: string]: unknown };
