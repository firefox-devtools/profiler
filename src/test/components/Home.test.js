/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import { Provider } from 'react-redux';

import { render } from 'firefox-profiler/test/fixtures/testing-library';
import { Home } from '../../components/app/Home';
import createStore from '../../app-logic/create-store';
import { mockWebChannel } from '../fixtures/mocks/web-channel';
import { fireFullClick } from '../fixtures/utils';

// ListOfPublishedProfiles depends on IDB and renders asynchronously, so we'll
// just test we want to render it, but otherwise test it more fully in a
// separate test file.
jest.mock('../../components/app/ListOfPublishedProfiles', () => ({
  ListOfPublishedProfiles: 'list-of-published-profiles',
}));

// Provide a mechanism to overwrite the navigator.userAgent, which can't be set.
const FIREFOX =
  'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.12; rv:55.0) Gecko/20100101 Firefox/55.0';
const SAFARI =
  'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/603.3.8 (KHTML, like Gecko) Version/10.1.2 Safari/603.3.8';
const CHROME =
  'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/603.3.8 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36';
let userAgent;

// Flow doesn't understand Object.defineProperty. Use the "any" type to use it anyway.
(Object.defineProperty: any)(window.navigator, 'userAgent', {
  get: () => userAgent,
});

describe('app/Home', function () {
  function setup(userAgentToConfigure: string) {
    userAgent = userAgentToConfigure;
    const renderResults = render(
      <Provider store={createStore()}>
        <Home specialMessage="This is a special message" />
      </Provider>
    );

    return { ...renderResults };
  }

  it('renders a button to enable the popup in Firefox', () => {
    const { container } = setup(FIREFOX);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the Chrome extension instructions for Chromium based browsers', () => {
    const { container } = setup(CHROME);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the information screen for other browsers', () => {
    const { container } = setup(SAFARI);
    expect(container.firstChild).toMatchSnapshot();
  });

  // This test's assertions are that it can find elements through getByTestId.
  // eslint-disable-next-line jest/expect-expect
  it('will switch to recording instructions when enabling the popup', async () => {
    const { triggerResponse, getLastRequestId } = mockWebChannel();
    const { findByTestId, getByText } = setup(FIREFOX);

    // Respond back from the browser that the menu button is not yet enabled.
    triggerResponse({
      type: 'SUCCESS_RESPONSE',
      requestId: getLastRequestId(),
      response: {
        menuButtonIsEnabled: false,
      },
    });
    await findByTestId('home-enable-popup-instructions');

    fireFullClick(getByText('Enable \u2068Firefox Profiler\u2069 Menu Button'));

    // Respond back from the browser that the menu button was enabled.
    triggerResponse({
      type: 'SUCCESS_RESPONSE',
      requestId: getLastRequestId(),
      response: undefined,
    });
    await findByTestId('home-record-instructions');
  });

  it('renders an error if the WebChannel is not available', async () => {
    // This simulates what happens if the profiler is run from a host which
    // is not the configured profiler base-url, or in an old Firefox version (<76)
    // which has WebChannels but no profiler WebChannel.

    const { listeners, triggerResponse } = mockWebChannel();

    // No one has asked anything to the WebChannel.
    expect(listeners).toHaveLength(0);

    const { findByText } = setup(FIREFOX);

    // There is an outstanding question to the WebChannel
    expect(listeners.length).toBeGreaterThan(0);

    // Respond from the browser that the WebChannel does not exist.
    jest.spyOn(console, 'error').mockImplementation(() => {});
    triggerResponse({
      errno: 2, // ERRNO_NO_SUCH_CHANNEL
      error: 'No such channel',
    });
    expect(console.error).toHaveBeenCalled();

    // The UI should update to include a note about the WebChannel being unavailable,
    // which is an async handler of the WebChannel message.
    const webChannelUnavailableMessage = await findByText(
      /This profiler instance was unable to connect to the WebChannel/
    );

    expect(webChannelUnavailableMessage).toMatchSnapshot();
  });
});
