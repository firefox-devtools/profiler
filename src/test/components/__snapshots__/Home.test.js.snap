// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/Home renders a button to enable the popup in Firefox 1`] = `
<div
  class="home"
>
  <main
    class="homeSection"
  >
    <header>
      <h1
        class="appHeader"
      >
        <span
          class="appHeaderSlogan"
        >
          <a
            class="appHeaderLink"
            href="/"
          >
            ⁨Firefox Profiler⁩
          </a>
           — 
          <span
            class="appHeaderSubtext"
          >
            Web app for ⁨Firefox⁩ performance analysis
          </span>
        </span>
        <a
          class="appHeaderGithubIcon"
          href="https://github.com/firefox-devtools/profiler"
          rel="noopener noreferrer"
          target="_blank"
          title="Go to our Git repository (this opens in a new window)"
        >
          <svg
            aria-label="github"
            class="octicon octicon-mark-github"
            height="22"
            version="1.1"
            viewBox="0 0 16 16"
            width="22"
          >
            <path
              d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0 0 16 8c0-4.42-3.58-8-8-8z"
              fill-rule="evenodd"
            />
          </svg>
        </a>
      </h1>
    </header>
    <div
      class="homeSpecialMessage"
    >
      This is a special message
    </div>
    <p>
      Capture a performance profile. Analyze it. Share it. Make the web faster.
    </p>
    <div
      class="homeInstructionsTransitionGroup"
    >
      <div
        class="homeInstructions"
        data-testid="home-enable-popup-instructions"
      >
        <img
          alt="screenshot of profiler.firefox.com"
          class="homeSectionScreenshot"
          src="test-file-stub"
        />
        <div>
          <button
            class="homeSectionButton"
            type="button"
          >
            <span
              class="homeSectionPlus"
            >
              +
            </span>
            Enable ⁨Firefox Profiler⁩ Menu Button
          </button>
          <a
            class="homeSectionButton"
            href="/docs/"
          >
            <span
              class="homeSectionDocsIcon"
            />
            Documentation
          </a>
          <p>
            Enable the profiler menu button to start recording a performance
profile in ⁨Firefox⁩, then analyze it and share it with profiler.firefox.com.
          </p>
          <p>
            You can also profile ⁨Firefox for Android⁩. For more
information, please consult this documentation:

            <a
              href="https://profiler.firefox.com/docs/#/./guide-profiling-android-directly-on-device?id=profiling-firefox-for-android-directly-on-device"
            >
              Profiling ⁨Firefox for Android⁩ directly on device
            </a>
            .
          </p>
        </div>
      </div>
    </div>
    <section
      class="homeAdditionalContent"
    >
      <h2
        class="homeAdditionalContentTitle protocol-display-xs"
      >
        Load existing profiles
      </h2>
      <section
        class="homeActions"
      >
        <p>
          You can 
          <strong>
            drag and drop
          </strong>
           a profile file here to load it, or:
        </p>
        <div
          class="homeSectionLoadProfile"
        >
          <div
            class="homeSectionActionButtons"
          >
            <input
              class="homeSectionUploadFromFileInput"
              type="file"
            />
            <button
              class="homeSectionButton"
              type="button"
            >
              Load a profile from file
            </button>
            <button
              aria-expanded="false"
              class="homeSectionButton"
              type="button"
            >
              Load a profile from a URL
            </button>
          </div>
        </div>
        <p>
          The ⁨Firefox Profiler⁩ can also import profiles from other profilers, such as

          <a
            href="https://profiler.firefox.com/docs/#/./guide-perf-profiling"
          >
            Linux perf
          </a>
          , 
          <a
            href="https://profiler.firefox.com/docs/#/./guide-android-profiling"
          >
            Android SimplePerf
          </a>
          , the
Chrome performance panel, 
          <a
            href="https://developer.android.com/studio/profile/cpu-profiler"
          >
            Android Studio
          </a>
          , or
any file using the 
          <a
            href="https://valgrind.org/docs/manual/dh-manual.html"
          >
            dhat format
          </a>
           or 
          <a
            href="https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview"
          >
            Google’s Trace Event
Format
          </a>
          . 
          <a
            href="https://github.com/firefox-devtools/profiler/blob/main/docs-developer/custom-importer.md"
          >
            Learn how to write your
own importer
          </a>
          .
        </p>
        <p>
          You can also compare recordings. 
          <a
            href="/compare/"
          >
            Open the comparing interface.
          </a>
        </p>
      </section>
      <section>
        <h2
          class="homeRecentUploadedRecordingsTitle protocol-display-xxs"
        >
          Your recent uploaded recordings
        </h2>
        <list-of-published-profiles
          limit="3"
          withactionbuttons="false"
        />
      </section>
    </section>
    <div
      class="dragAndDropOverlayWrapper"
    >
      <div
        class="dragAndDropOverlay"
      >
        Drop a saved profile here
      </div>
    </div>
  </main>
</div>
`;

exports[`app/Home renders an error if the WebChannel is not available 1`] = `
<p>
  This profiler instance was unable to connect to the WebChannel. This usually means that it’s
running on a different host from the one that is specified in the preference

  <code>
    devtools.performance.recording.ui-base-url
  </code>
  . If you would like to capture new
profiles with this instance, and give it programmatic control of the profiler menu button,
you can go to 
  <code>
    about:config
  </code>
   and change the preference.
</p>
`;

exports[`app/Home renders the Chrome extension instructions for Chromium based browsers 1`] = `
<div
  class="home"
>
  <main
    class="homeSection"
  >
    <header>
      <h1
        class="appHeader"
      >
        <span
          class="appHeaderSlogan"
        >
          <a
            class="appHeaderLink"
            href="/"
          >
            ⁨Firefox Profiler⁩
          </a>
           — 
          <span
            class="appHeaderSubtext"
          >
            Web app for ⁨Firefox⁩ performance analysis
          </span>
        </span>
        <a
          class="appHeaderGithubIcon"
          href="https://github.com/firefox-devtools/profiler"
          rel="noopener noreferrer"
          target="_blank"
          title="Go to our Git repository (this opens in a new window)"
        >
          <svg
            aria-label="github"
            class="octicon octicon-mark-github"
            height="22"
            version="1.1"
            viewBox="0 0 16 16"
            width="22"
          >
            <path
              d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0 0 16 8c0-4.42-3.58-8-8-8z"
              fill-rule="evenodd"
            />
          </svg>
        </a>
      </h1>
    </header>
    <div
      class="homeSpecialMessage"
    >
      This is a special message
    </div>
    <p>
      Capture a performance profile. Analyze it. Share it. Make the web faster.
    </p>
    <div
      class="homeInstructionsTransitionGroup"
    >
      <div
        class="homeInstructions"
        data-testid="home-enable-popup-instructions"
      >
        <img
          alt="screenshot of profiler.firefox.com"
          class="homeSectionScreenshot"
          src="test-file-stub"
        />
        <div>
          <a
            class="homeSectionButton"
            href="https://chromewebstore.google.com/detail/firefox-profiler/ljmahpnflmbkgaipnfbpgjipcnahlghn"
            rel="noreferrer"
            target="_blank"
          >
            <span
              class="homeSectionPlus"
            >
              +
            </span>
            Install the Chrome extension
          </a>
          <a
            class="homeSectionButton"
            href="/docs/"
          >
            <span
              class="homeSectionDocsIcon"
            />
            Documentation
          </a>
          <p>
            Use the 
            <a
              href="https://chromewebstore.google.com/detail/firefox-profiler/ljmahpnflmbkgaipnfbpgjipcnahlghn"
              rel="noreferrer"
              target="_blank"
            >
              ⁨Firefox Profiler⁩ extension for Chrome
            </a>
            
to capture performance profiles in Chrome and analyze them in the
⁨Firefox Profiler⁩. Install the extension from the Chrome Web Store.
          </p>
          <p>
            Once installed, use the extension’s
toolbar icon or the shortcuts to start and stop profiling. You can also
export profiles and load them here for detailed analysis.
          </p>
          <div>
            <p>
              <kbd>
                Ctrl
              </kbd>
              +
              <kbd>
                Shift
              </kbd>
              +
              <kbd>
                1
              </kbd>
               
              Stop and start profiling
            </p>
            <p>
              <kbd>
                Ctrl
              </kbd>
              +
              <kbd>
                Shift
              </kbd>
              +
              <kbd>
                2
              </kbd>
               
              Capture and load profile
            </p>
          </div>
        </div>
      </div>
    </div>
    <section
      class="homeAdditionalContent"
    >
      <h2
        class="homeAdditionalContentTitle protocol-display-xs"
      >
        Load existing profiles
      </h2>
      <section
        class="homeActions"
      >
        <p>
          You can 
          <strong>
            drag and drop
          </strong>
           a profile file here to load it, or:
        </p>
        <div
          class="homeSectionLoadProfile"
        >
          <div
            class="homeSectionActionButtons"
          >
            <input
              class="homeSectionUploadFromFileInput"
              type="file"
            />
            <button
              class="homeSectionButton"
              type="button"
            >
              Load a profile from file
            </button>
            <button
              aria-expanded="false"
              class="homeSectionButton"
              type="button"
            >
              Load a profile from a URL
            </button>
          </div>
        </div>
        <p>
          The ⁨Firefox Profiler⁩ can also import profiles from other profilers, such as

          <a
            href="https://profiler.firefox.com/docs/#/./guide-perf-profiling"
          >
            Linux perf
          </a>
          , 
          <a
            href="https://profiler.firefox.com/docs/#/./guide-android-profiling"
          >
            Android SimplePerf
          </a>
          , the
Chrome performance panel, 
          <a
            href="https://developer.android.com/studio/profile/cpu-profiler"
          >
            Android Studio
          </a>
          , or
any file using the 
          <a
            href="https://valgrind.org/docs/manual/dh-manual.html"
          >
            dhat format
          </a>
           or 
          <a
            href="https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview"
          >
            Google’s Trace Event
Format
          </a>
          . 
          <a
            href="https://github.com/firefox-devtools/profiler/blob/main/docs-developer/custom-importer.md"
          >
            Learn how to write your
own importer
          </a>
          .
        </p>
        <p>
          You can also compare recordings. 
          <a
            href="/compare/"
          >
            Open the comparing interface.
          </a>
        </p>
      </section>
      <section>
        <h2
          class="homeRecentUploadedRecordingsTitle protocol-display-xxs"
        >
          Your recent uploaded recordings
        </h2>
        <list-of-published-profiles
          limit="3"
          withactionbuttons="false"
        />
      </section>
    </section>
    <div
      class="dragAndDropOverlayWrapper"
    >
      <div
        class="dragAndDropOverlay"
      >
        Drop a saved profile here
      </div>
    </div>
  </main>
</div>
`;

exports[`app/Home renders the information screen for other browsers 1`] = `
<div
  class="home"
>
  <main
    class="homeSection"
  >
    <header>
      <h1
        class="appHeader"
      >
        <span
          class="appHeaderSlogan"
        >
          <a
            class="appHeaderLink"
            href="/"
          >
            ⁨Firefox Profiler⁩
          </a>
           — 
          <span
            class="appHeaderSubtext"
          >
            Web app for ⁨Firefox⁩ performance analysis
          </span>
        </span>
        <a
          class="appHeaderGithubIcon"
          href="https://github.com/firefox-devtools/profiler"
          rel="noopener noreferrer"
          target="_blank"
          title="Go to our Git repository (this opens in a new window)"
        >
          <svg
            aria-label="github"
            class="octicon octicon-mark-github"
            height="22"
            version="1.1"
            viewBox="0 0 16 16"
            width="22"
          >
            <path
              d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0 0 16 8c0-4.42-3.58-8-8-8z"
              fill-rule="evenodd"
            />
          </svg>
        </a>
      </h1>
    </header>
    <div
      class="homeSpecialMessage"
    >
      This is a special message
    </div>
    <p>
      Capture a performance profile. Analyze it. Share it. Make the web faster.
    </p>
    <div
      class="homeInstructionsTransitionGroup"
    >
      <div
        class="homeInstructions"
        data-testid="home-other-browser-instructions"
      >
        <img
          alt="screenshot of profiler.firefox.com"
          class="homeSectionScreenshot"
          src="test-file-stub"
        />
        <div>
          <a
            class="homeSectionButton"
            href="/docs/"
          >
            <span
              class="homeSectionDocsIcon"
            />
            Documentation
          </a>
          <p>
            Recording performance profiles requires 
            <a
              href="https://www.mozilla.org/en-US/firefox/new/"
            >
              ⁨Firefox⁩
            </a>
            .
However, existing profiles can be viewed in any modern browser.
          </p>
          <p>
            You can also profile ⁨Firefox for Android⁩. For more
information, please consult this documentation:

            <a
              href="https://profiler.firefox.com/docs/#/./guide-profiling-android-directly-on-device?id=profiling-firefox-for-android-directly-on-device"
            >
              Profiling ⁨Firefox for Android⁩ directly on device
            </a>
            .
          </p>
        </div>
      </div>
    </div>
    <section
      class="homeAdditionalContent"
    >
      <h2
        class="homeAdditionalContentTitle protocol-display-xs"
      >
        Load existing profiles
      </h2>
      <section
        class="homeActions"
      >
        <p>
          You can 
          <strong>
            drag and drop
          </strong>
           a profile file here to load it, or:
        </p>
        <div
          class="homeSectionLoadProfile"
        >
          <div
            class="homeSectionActionButtons"
          >
            <input
              class="homeSectionUploadFromFileInput"
              type="file"
            />
            <button
              class="homeSectionButton"
              type="button"
            >
              Load a profile from file
            </button>
            <button
              aria-expanded="false"
              class="homeSectionButton"
              type="button"
            >
              Load a profile from a URL
            </button>
          </div>
        </div>
        <p>
          The ⁨Firefox Profiler⁩ can also import profiles from other profilers, such as

          <a
            href="https://profiler.firefox.com/docs/#/./guide-perf-profiling"
          >
            Linux perf
          </a>
          , 
          <a
            href="https://profiler.firefox.com/docs/#/./guide-android-profiling"
          >
            Android SimplePerf
          </a>
          , the
Chrome performance panel, 
          <a
            href="https://developer.android.com/studio/profile/cpu-profiler"
          >
            Android Studio
          </a>
          , or
any file using the 
          <a
            href="https://valgrind.org/docs/manual/dh-manual.html"
          >
            dhat format
          </a>
           or 
          <a
            href="https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview"
          >
            Google’s Trace Event
Format
          </a>
          . 
          <a
            href="https://github.com/firefox-devtools/profiler/blob/main/docs-developer/custom-importer.md"
          >
            Learn how to write your
own importer
          </a>
          .
        </p>
        <p>
          You can also compare recordings. 
          <a
            href="/compare/"
          >
            Open the comparing interface.
          </a>
        </p>
      </section>
      <section>
        <h2
          class="homeRecentUploadedRecordingsTitle protocol-display-xxs"
        >
          Your recent uploaded recordings
        </h2>
        <list-of-published-profiles
          limit="3"
          withactionbuttons="false"
        />
      </section>
    </section>
    <div
      class="dragAndDropOverlayWrapper"
    >
      <div
        class="dragAndDropOverlay"
      >
        Drop a saved profile here
      </div>
    </div>
  </main>
</div>
`;
